dora/src/bytecode/writer.rs

```diff
@@ -965,34 +972,62 @@ impl BytecodeWriter {
         fid: FieldId,
     ) {
         let values = [
-            inst as u32,
             r1.to_usize() as u32,
             r2.to_usize() as u32,
             cid.to_usize() as u32,
             fid.to_usize() as u32,
         ];
-        self.emit_values(&values);
+        self.emit_values(inst, &values);
     }
 
     fn emit_load_global(&mut self, inst: BytecodeOpcode, r1: Register, gid: GlobalId) {
-        let values = [inst as u32, r1.to_usize() as u32, gid.to_usize() as u32];
-        self.emit_values(&values);
+        let values = [r1.to_usize() as u32, gid.to_usize() as u32];
+        self.emit_values(inst, &values);
+    }
+
+    pub fn set_position(&mut self, pos: Position) {
+        self.position = Some(pos);
     }
 
     fn emit_op(&mut self, inst: BytecodeOpcode) {
-        let values = [inst as u32];
-        self.emit_values(&values);
+        let values = [];
+        self.emit_values(inst, &values);
     }
 
-    fn emit_values(&mut self, values: &[u32]) {
+    fn emit_position(&mut self) {
+        let offset = self.code.len() as u32;
+        if self.position.is_none() {
+            return;
+        }
+
+        let position = self.position.unwrap();
+        let last_position = self.positions.last().map(|(_, p)| p);
+
+        if let Some(last_position) = last_position {
+            if *last_position == position {
+                return;
+            }
+        }
+
+        self.positions.push((offset, position));
+        self.position = None;
+    }
+
+    fn emit_values(&mut self, op: BytecodeOpcode, values: &[u32]) {
+        if op.need_position() {
+            self.emit_position();
+        }
+
         let is_wide = values.iter().any(|&val| val > u8::max_value() as u32);
```

Comment:
I think this needs to change now, since it is now not checking the opcode anymore. This should now also check whether `op` fits in a single byte. I know that Bytecodes currently always fit in one byte, but we already have around 160 bytecodes and might soon not fit into a byte anymore.